modules/fedimint-ln-server/src/lib.rs

```diff
@@ -268,35 +278,96 @@ impl ServerModule for Lightning {
         &self,
         dbtx: &mut ModuleDatabaseTransaction<'_, ModuleInstanceId>,
     ) {
-        if !self.consensus_proposal(dbtx).await.forces_new_epoch() {
-            std::future::pending().await
+        while !self.consensus_proposal(dbtx).await.forces_new_epoch() {
+            // FIXME: remove after modularization finishes
+            #[cfg(not(target_family = "wasm"))]
+            sleep(Duration::from_millis(1000)).await;
         }
     }
 
     async fn consensus_proposal(
         &self,
         dbtx: &mut ModuleDatabaseTransaction<'_, ModuleInstanceId>,
     ) -> ConsensusProposal<LightningConsensusItem> {
-        ConsensusProposal::new_auto_trigger(
-            dbtx.find_by_prefix(&ProposeDecryptionShareKeyPrefix)
-                .await
-                .map(
-                    |(ProposeDecryptionShareKey(contract_id), share)| LightningConsensusItem {
-                        contract_id,
-                        share,
-                    },
-                )
-                .collect::<Vec<LightningConsensusItem>>()
-                .await,
-        )
+        let round_ci = LightningConsensusItem::RoundConsensus(RoundConsensusItem {
+            clock_time: fedimint_core::time::now()
+                .duration_since(UNIX_EPOCH)
+                .expect("Error getting clock time")
+                .as_secs(),
+        });
+
+        let items = dbtx
+            .find_by_prefix(&ProposeDecryptionShareKeyPrefix)
+            .await
+            .map(|(ProposeDecryptionShareKey(contract_id), share)| {
+                LightningConsensusItem::DecryptionShare(DecryptionShareCI { contract_id, share })
+            })
+            .chain(stream::once(async { round_ci }))
+            .collect::<Vec<LightningConsensusItem>>()
+            .await;
+
+        // We force new epochs with new clock time only if invoice expires to avoid
+        // continually producing epochs
+        let mut outgoing_contracts: Vec<OutgoingContract> = dbtx
+            .find_by_prefix(&ContractKeyPrefix)
+            .await
+            .filter_map(|(_, contract)| async move {
+                if let FundedContract::Outgoing(outgoing_contract) = contract.contract {
+                    Some(outgoing_contract)
+                } else {
+                    None
+                }
+            })
+            .collect()
+            .await;
+
+        outgoing_contracts
+            .sort_by_key(|contract| contract.invoice.timestamp() + contract.invoice.expiry_time());
+
+        // We force new epochs only if invoice expires, or we have decryption share
+        // (more than just round_ci item)
+        if 1 < items.len()
+            || (outgoing_contracts
```

Comment:
Looks better!  Could we just `.filter(|c| c.invoice.is_expired())` on `outgoing_contracts` and then check if `len() > 0`?